File: src/com/l2kt/gameserver/network/serverpackets/L2GameServerPacket.java
package com.l2kt.gameserver.network.serverpackets;

import com.l2kt.Config;
import com.l2kt.commons.logging.CLogger;
import com.l2kt.commons.mmocore.SendablePacket;

import com.l2kt.gameserver.network.L2GameClient;

public abstract class L2GameServerPacket extends SendablePacket<L2GameClient>
{
	protected static final CLogger LOGGER = new CLogger(L2GameServerPacket.class.getName());
	
	protected abstract void writeImpl();
	
	@Override
	protected void write()
	{
		if (Config.PACKET_HANDLER_DEBUG)
			LOGGER.info(getType());
		
		try
		{
			writeImpl();
		}
		catch (Throwable t)
		{
			LOGGER.error("Failed writing {} for {}. ", t, getType(), getClient().toString());
		}
	}
	
	public void runImpl()
	{
	}
	
	public String getType()
	{
		return "[S] " + getClass().getSimpleName();
	}
}

File: src/com/l2kt/gameserver/network/serverpackets/L2GameServerPacket.kt
package com.l2kt.gameserver.network.serverpackets

import com.l2kt.Config
import com.l2kt.commons.logging.CLogger
import com.l2kt.commons.mmocore.SendablePacket

import com.l2kt.gameserver.network.L2GameClient

abstract class L2GameServerPacket : SendablePacket<L2GameClient>() {

    val type: String
        get() = "[S] " + javaClass.simpleName

    protected abstract fun writeImpl()

    override fun write() {
        if (Config.PACKET_HANDLER_DEBUG)
            LOGGER.info(type)

        try {
            writeImpl()
        } catch (t: Throwable) {
            LOGGER.error("Failed writing {} for {}. ", t, type, client.toString())
        }

    }

    open fun runImpl() {}

    companion object {
        val LOGGER = CLogger(L2GameServerPacket::class.java.name)
    }
}